src/us/kbase/workspace/database/provenance/CreditMetadata.java

```diff
@@ -0,0 +1,465 @@
+			return this;
+		}
+
+		/**
+		 * Sets the license for the resource; can be a string like 'Apache 2.0' or an
+		 * URL-like string; URL-like strings will be checked for well-formedness.
+		 * For these purposes, a license string that starts with a series of letters
+		 * followed by one or more colons and one or more slashes (e.g. "https://") is
+		 * considered URL-like.
+		 *
+		 * @param license
+		 *                license as a string
+		 * @return this builder
+		 */
+		public Builder withLicense(final String license) {
+			this.license = Common.processString(license);
+			return this;
+		}
+
+		/**
+		 * Sets the version.
+		 *
+		 * Valid credit metadata objects require either a version or one or more dates.
+		 *
+		 * @param withVersion
+		 *                version string
+		 * @return this builder
+		 */
+		public Builder withVersion(final String version) {
+			this.version = Common.processString(version);
+			return this;
+		}
+
+		/**
+		 * Sets the dates of various resource life cycle events.
+		 *
+		 * Valid credit metadata objects require either a version or one or more dates.
+		 *
+		 * @param withDates
+		 *                list of {@link EventDate}s.
+		 * @return this builder
+		 */
+		public Builder withDates(final List<EventDate> dates) {
+			this.dates = Common.dedupeSimpleList(dates);
+			return this;
+		}
+
+		/**
+		 * Sets the funding source(s).
+		 *
+		 * @param withFunding
+		 *                list of {@link FundingReference}s.
+		 * @return this builder
+		 */
+		public Builder withFunding(final List<FundingReference> funding) {
+			this.funding = Common.dedupeSimpleList(funding);
+			return this;
+		}
+
+		/**
+		 * Sets the related identifiers.
+		 *
+		 * @param withRelatedIdentifiers
+		 *                list of {@link PermanentID}s.
+		 * @return this builder
+		 */
+		public Builder withRelatedIdentifiers(final List<PermanentID> relatedIdentifiers) {
+			this.relatedIdentifiers = Common.dedupeSimpleList(relatedIdentifiers);
+			return this;
+		}
+
+		/**
+		 * Build the {@link CreditMetadata}.
+		 *
+		 * In addition to the required fields set in getBuilder(), credit metadata
+		 * requires some sort of versioning information -- either an explicit version
+		 * identifier or one or more dates.
+		 *
+		 * @return the credit metadata.
+		 */
+		public CreditMetadata build() {
+
+			// ensure there is either a version or a non-empty list of dates
+			if (version == null && (dates == null || dates.isEmpty())) {
+				errorList.add("must provide either 'version' or one or more " +
+						"'dates', ideally indicating when the resource " +
+						"was published or when it was last updated");
+			}
+
+			// If the license looks like a URL, ensure it's properly formatted and seems
+			// valid. For these purposes, a license string that starts with a series
+			// of letters followed by one or more colons and one or more slashes is
+			// considered URL-like.
+			Pattern pattern = Pattern.compile("^[a-zA-Z]+:+/+");
```

Comment:
Make this static so it's only compiled once

Reply:
I see it's now final, but it'll still be compiled every call - I mean to make it a static class variable so it's only compiled once the first time the class is imported

Reply:
Doh, got it.

Reply:
:+1: but see https://github.com/kbase/workspace_deluxe/pull/664#discussion_r1113667872